### DISCUSSION TITLE
How to store version number of a program?

### DISCUSSION THREAD

**Discussion by 2-4601** [twpayne/chezmoi#2564] [Q&A]
I use `mpv` media player on two different computers. Because they have different versions of `mpv`, I have to make some adjustments and `chezmoi` is perfect for this job.
For example, in my `.chezmoiexternal.toml` I have:
```
{{ $mpv_version := output "sh" "-c" "mpv --version | sed -n 's/^mpv \\([.0-9]\\{1,\\}\\).*$/\\1/p'" | trim }}
{{/* uosc requires at least mpv version 0.33.0 */}}
{{ if semverCompare ">=0.33.0" $mpv_version }}
[".config/mpv/"]
    type = "archive"
    url = "https://github.com/tomasklaen/uosc/releases/download/{{ (gitHubLatestRelease "tomasklaen/uosc").TagName }}/uosc.zip"
    refreshPeriod = "168h"
{{ end }}
```

Then in my `mpv.conf.tmpl` I need the version number again because older versions don't support certain settings:
```
{{ $mpv_version := output "sh" "-c" "mpv --version | sed -n 's/^mpv \\([.0-9]\\{1,\\}\\).*$/\\1/p'" | trim }}
{{ if semverCompare ">=0.34.1" $mpv_version }}
watch-later-options-remove=volume
{{ end }}
```

This setup works, but I'm just wondering if there's a way to store `$mpv_version` only once when I run `chezmoi apply` and use its value in different files? Storing it in the `chezmoi` config file requires running an explicit init command and I would have to remember to do that when `mpv` is updated to newer versions.

**Comment by halostatue** [comment]
Each template is run isolated and as far as I can tell has no way of modifying the chezmoi runtime data.

You could consider writing the comparison to a temporary file which then gets deleted using a `run_after` script.

**Comment by bradenhilton** [comment]
If you use chezmoi to store the version number somewhere, and read it inside the source state, it will always be behind the source state.

Take the following for example:

1. You use a `run_onchange_` template script to detect a new version of mpv and print its version to a file, e.g.:
    ```
    # {{ output "sh" "-c" "mpv --version | sed -n 's/^mpv \\([.0-9]\\{1,\\}\\).*$/\\1/p'" | trim | sha256sum }}
    mpv --version | sed -n 's/^mpv \\([.0-9]\\{1,\\}\\).*$/\\1/p' > {{ joinPath .chezmoi.homeDir "mpv_version" }}
    ```
2. You read this file in your templates:
    ```
    {{ include (joinPath .chezmoi.homeDir "mpv_version") }}
    ```

If you set the contents of the `mpv_version` file to `0` and run `chezmoi diff`, you will see that it reads `0` in your templates. Only after running `chezmoi apply` will the `mpv_version` file be updated.

**Not to mention that this [violates the assumption that the source and destination states do not change while chezmoi is running](https://www.chezmoi.io/reference/application-order/).**

Do you have any particular concerns with your current implementation? It looks fine to me, you're always reading the version when you need it, and it's always going to be correct.

**Reply by 2-4601** [reply]
> Do you have any particular concerns with your current implementation?

No, other than that the call to parse the version number is just very verbose and ugly :) I have to use it in three different files currently. I was just wondering because `chezmoi` has so many features that there could be a more elegant way to solve what I'm doing.

---------------

After reading your example I found out about [`includeTemplate`](https://www.chezmoi.io/reference/templates/functions/includeTemplate/). It seems to solve my issue. For example:

`.chezmoitemplates/mpv_version`
```
{{- output "sh" "-c" "mpv --version | sed -n 's/^mpv \\([.0-9]\\{1,\\}\\).*$/\\1/p'" | trim -}}
```

`mpv.conf.tmpl`
```
{{- $mpv_version := includeTemplate "mpv_version" -}}
{{ if semverCompare ">=0.34.1" $mpv_version }}
watch-later-options-remove=volume
{{ end }}
```

I'm not sure if this violates _"the assumption that the source and destination states do not change while chezmoi is running"_? Or if there are other issues with it?

**Reply by halostatue** [reply]
It doesn’t violate the assumption any more than the use of `lookPath` does. You might be able to simplify your call sites a little further:

```
# .chezmoitemplates/mpv_version_compare
{{- $mpv_version := output "sh" "-c" "mpv --version | sed -n 's/^mpv \\([.0-9]\\{1,\\}\\).*$/\\1/p'" | trim -}}
{{- semverCompare . $mpv_version -}}
```

```
# mpv.conf.tmpl
{{ if includeTemplate "mpv_version_compare" ">=0.34.1" }}
watch-later-options-remove=volume
{{ end }}
```

This is untested, so it might not work.

**Reply by 2-4601** [reply]
Yeah, that almost works... but as the non-string types seem to get lost, everything needs very explicit conversion:

`.chezmoitemplates/mpv_version_compare`:
```
{{- $mpv_version := output "sh" "-c" "mpv --version | sed -n 's/^mpv \\([.0-9]\\{1,\\}\\).*$/\\1/p'" | trim -}}
{{- semverCompare . $mpv_version | int -}}
```

`mpv.conf.tmpl`:
```
{{ if includeTemplate "mpv_version_compare" ">=0.34.1" | int }}
watch-later-options-remove=volume
{{ end }}
```

**Reply by bradenhilton** [reply]
Your `mpv_version` template is easier to reason about than the `mpv_version_compare` template, IMO.

Also, if you aren't using the `$mpv_version` variable anywhere else, I think you can just simplify it to 

```
{{ if semverCompare ">=0.34.1" (includeTemplate "mpv_version") }}
```
